Model/Service/Sync/AsyncBulkConsumer.php

```diff
@@ -0,0 +1,140 @@
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @author Nosto Solutions Ltd <contact@nosto.com>
+ * @copyright 2019 Nosto Solutions Ltd
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD 3-Clause
+ *
+ */
+
+namespace Nosto\Tagging\Model\Service\Sync;
+
+use Exception;
+use Magento\Framework\EntityManager\EntityManager;
+use Magento\Framework\Json\Helper\Data as JsonHelper;
+use Nosto\Tagging\Logger\Logger;
+use Nosto\Tagging\Model\Product\Index\IndexRepository;
+use Nosto\Tagging\Model\Service\Sync\SyncService as NostoSyncService;
+use Nosto\Tagging\Helper\Scope as NostoScopeHelper;
+
+/**
+ * Asynchronous Bulk Consumer
+ *
+ * Class AsyncBulkConsumer
+ */
+class AsyncBulkConsumer
+{
+    /** @var Logger */
+    private $logger;
+
+    /** @var JsonHelper */
+    private $jsonHelper;
+
+    /** @var IndexRepository */
+    private $indexRepository;
+
+    /** @var NostoSyncService */
+    private $nostoSyncService;
+
+    /** @var NostoScopeHelper */
+    private $nostoScopeHelper;
+
+    /** @var EntityManager */
+    private $entityManager;
+
+    /**
+     * AsyncBulkConsumer constructor.
+     *
+     * @param Logger $logger
+     * @param JsonHelper $jsonHelper
+     * @param IndexRepository $indexRepository
+     * @param NostoSyncService $nostoSyncService
+     * @param NostoScopeHelper $nostoScopeHelper
+     * @param EntityManager $entityManager
+     */
+    public function __construct(
+        Logger $logger,
+        JsonHelper $jsonHelper,
+        IndexRepository $indexRepository,
+        NostoSyncService $nostoSyncService,
+        NostoScopeHelper $nostoScopeHelper,
+        EntityManager $entityManager
+    ) {
+        $this->logger = $logger;
+        $this->jsonHelper = $jsonHelper;
+        $this->indexRepository = $indexRepository;
+        $this->nostoSyncService = $nostoSyncService;
+        $this->nostoScopeHelper = $nostoScopeHelper;
+        $this->entityManager = $entityManager;
+    }
+
+    /**
+     * Processing operation for product sync
+     *
+     * @param array|\Magento\AsynchronousOperations\Api\Data\OperationInterface $operation
+     * @return void
+     * @throws Exception
+     * @suppress PhanUndeclaredClassConstant
+     */
+    public function processOperation($operation)
+    {
+        $errorCode = null;
+        $message = null;
+        if (!is_array($operation)) {
```

Comment:
Should we first check if `$operation` is an array. If not check that the type is `OperationInterface` and if that fails we should throw an exception.